python/tests/test_combinatorics.py

```diff
@@ -0,0 +1,668 @@
+    n = t.num_leaves()
+    k = t.left.num_leaves()
+    total = comb.nCr(n, k) * L_tree(t.left) * L_tree(t.right)
+
+    if t.is_symmetrical():
+        assert total % 2 == 0
+        return total // 2
+    return total
+
+
+class Leaf:
+    def __init__(self, val):
+        self.val = val
+
+    def shape_rank(self):
+        return 0
+
+    def label_rank(self):
+        return 0
+
+    def label_unrank(self, rank, labels=None):
+        if labels is None:
+            labels = [0]
+
+        assert rank == 0
+        assert len(labels) == 1
+        self.val = labels[0]
+
+    def __str__(self):
+        return str(self.val)
+
+    def num_leaves(self):
+        return 1
+
+    def leaves(self):
+        return [self.val]
+
+    def height(self):
+        return 0
+
+    def shape_equal(self, other):
+        return isinstance(other, Leaf)
+
+    def equal(self, other):
+        return isinstance(other, Leaf) and other.val == self.val
+
+    def is_canonical(self):
+        return True
+
+    def canonical(self):
+        return self
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        return Leaf(tree.root)
+
+    def to_tsk_tree(self):
+        tables = tskit.TableCollection(1)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+        return tables.tree_sequence().first()
+
+
+class Tree:
+    """
+    A full, labeled, rooted binary tree. Isomorphic trees (equivalent up
+    to swapping sibling nodes) are considered equivalent.
+    """
+
+    def __init__(self, left, right):
+        self.left = left
+        self.right = right
+        self._num_leaves = left.num_leaves() + right.num_leaves()
+        self._cached_rank = None
+
+    def num_leaves(self):
+        return self._num_leaves
+
+    ########################################################################
+    # Generators
+    ########################################################################
+
+    @staticmethod
+    def all_labeled_trees(n):
+        """
+        Generate all unordered, leaf-labeled trees with n leaves.
+        """
+        for tree in Tree.all_unordered_trees(n):
+            yield from Tree.all_labelings(tree)
+
+    @staticmethod
+    def all_unordered_trees(n):
+        """
+        Generate all unordered trees of n leaves.
+
+        Let S(k) be the set of all unordered, unlabeled trees of k leaves.
+
+        A tree with n leaves can be obtained by joining a tree with k < n
+        leaves with a tree of n - k leaves for all 1 < k < n - 1.
+        This, however, results in distinct trees that are isomorphic to each
+        other since you could swap the left and right subtrees to get an already
+        existing tree.
+
+        To eliminate this consideration moving forward, we say a tree
+        T = T_l ^ T_r is in its canonical orientation, which we will call
+        left-heavy, if num_leaves(T_l) >= num_leaves(T_r), and both T_l and T_r
+        are themselves left-heavy. A leaf is trivially canonically oriented.
```

Comment:
This is a bit problematic because tskit orders edges by putting the child with the smaller ID on the left. Since all samples are smaller than internal nodes, tskit will naturally make the trees "right-heavy" if e.g. a node has one child that's a leaf and one that's an internal node. It doesn't matter for any of the combinatorics, and we don't expect naturally created tskit trees to be canonical, but it could be confusing if you were unranking trees and printing them.

Reply:
> Since all samples are smaller than internal nodes,

This isn't necessarily true, as sample IDs can be anything. It's just that the canonical (post simplify) version of a tree sequence always has samples as 0 to n.

Reply:
If we go with the approach of each node having a list of children, we can make the canonical form listing the children in nondecreasing order of subtended leaves. This seems quite natural to me and would match up with tskit.